Source/JavaScriptCore/runtime/JSGlobalObject.cpp

```diff
@@ -2505,19 +2498,46 @@ void JSGlobalObject::tryInstallSpeciesWatchpoint(JSObject* prototype, JSObject*
     PropertySlot speciesSlot(constructor, PropertySlot::InternalMethodType::VMInquiry, &vm);
     constructor->getOwnPropertySlot(constructor, this, vm.propertyNames->speciesSymbol, speciesSlot);
     scope.assertNoException();
-    if (speciesSlot.slotBase() != constructor
-        || !speciesSlot.isCacheableGetter()
-        || speciesSlot.getterSetter() != speciesGetterSetter()) {
-        invalidateWatchpoint();
-        return;
+    switch (speciesEmptyMode) {
+    case SpeciesEmpty::Yes: {
+        if (!speciesSlot.isUnset()) {
+            invalidateWatchpoint();
+            return;
+        }
+        break;
+    }
+    case SpeciesEmpty::No: {
+        if (speciesSlot.slotBase() != constructor
+            || !speciesSlot.isCacheableGetter()
+            || speciesSlot.getterSetter() != speciesGetterSetter()) {
+            invalidateWatchpoint();
+            return;
+        }
+        break;
+    }
     }
 
     // Now we need to setup the watchpoints to make sure these conditions remain valid.
+
     prototypeStructure->startWatchingPropertyForReplacements(vm, constructorSlot.cachedOffset());
-    constructorStructure->startWatchingPropertyForReplacements(vm, speciesSlot.cachedOffset());
+    switch (speciesEmptyMode) {
+    case SpeciesEmpty::Yes:
+        break;
+    case SpeciesEmpty::No:
+        constructorStructure->startWatchingPropertyForReplacements(vm, speciesSlot.cachedOffset());
+        break;
+    }
 
     ObjectPropertyCondition constructorCondition = ObjectPropertyCondition::equivalence(vm, prototype, prototype, vm.propertyNames->constructor.impl(), constructor);
-    ObjectPropertyCondition speciesCondition = ObjectPropertyCondition::equivalence(vm, prototype, constructor, vm.propertyNames->speciesSymbol.impl(), speciesGetterSetter());
+    ObjectPropertyCondition speciesCondition;
+    switch (speciesEmptyMode) {
+    case SpeciesEmpty::No:
+        speciesCondition = ObjectPropertyCondition::equivalence(vm, prototype, constructor, vm.propertyNames->speciesSymbol.impl(), speciesGetterSetter());
+        break;
+    case SpeciesEmpty::Yes:
+        speciesCondition = ObjectPropertyCondition::absence(vm, prototype, constructor, vm.propertyNames->speciesSymbol.impl(), jsDynamicCast<JSObject*>(constructor->getPrototypeDirect()));
+        break;
```

Comment:
Why is the `prototype` the `owner`? Shouldn't it be the global object?

Reply:
Actually, this is not so important since we do not need write-barrier. For now, just using `this`, but in the future, it should be `nullptr` (But I don't change it in this patch since I'm not planning to change all the other things too).